src/rdkafka_broker.c

```diff
@@ -1567,6 +1569,111 @@ int rd_kafka_socket_cb_generic (int domain, int type, int protocol,
 }
 
 
+
+/**
+ * @brief Update the reconnect backoff.
+ *        Should be called when a connection is made.
+ *
+ * @locality broker thread
+ * @locks none
+ */
+static void
+rd_kafka_broker_update_reconnect_backoff (rd_kafka_broker_t *rkb,
+                                          const rd_kafka_conf_t *conf,
+                                          rd_ts_t now) {
+        int backoff;
+
+        /* If last connection attempt was more than reconnect.backoff.max.ms
+         * ago, reset the reconnect backoff to the initial
+         * reconnect.backoff.ms value.
+         * Or if the current backoff increment has reached the max value. */
+        if (rkb->rkb_ts_reconnect + (conf->reconnect_backoff_max_ms * 1000) <
+            now ||
+            rkb->rkb_reconnect_backoff_ms >= conf->reconnect_backoff_max_ms)
+                rkb->rkb_reconnect_backoff_ms = conf->reconnect_backoff_ms;
+
+        /* Apply +-20% jitter to next backoff. */
+        backoff = rd_jitter((int)(float)rkb->rkb_reconnect_backoff_ms * 0.8,
+                            (int)(float)rkb->rkb_reconnect_backoff_ms * 1.2);
+
+         /* Cap to reconnect.backoff.max.ms. */
+        backoff = RD_MIN(backoff, conf->reconnect_backoff_max_ms);
+
+        /* Set time of next reconnect */
+        rkb->rkb_ts_reconnect = now + (backoff * 1000);
+        rkb->rkb_reconnect_backoff_ms *= 2;
+}
+
+
+/**
+ * @brief Calculate time until next reconnect attempt.
+ *
+ * @returns the number of milliseconds to the next connection attempt, or 0
+ *          if immediate.
+ * @locality broker thread
+ * @locks none
+ */
+
+static RD_INLINE int
+rd_kafka_broker_reconnect_backoff (const rd_kafka_broker_t *rkb,
+                                   rd_ts_t now) {
+        rd_ts_t remains;
+
+        if (unlikely(rkb->rkb_ts_reconnect == 0))
+                return 0; /* immediate */
+
+        remains = rkb->rkb_ts_reconnect - now;
+        if (remains <= 0)
+                return 0; /* immediate */
+
+        return (int)(remains / 1000);
```

Comment:
Can we avoid casting this to int, or at least check for overflow?

Reply:
Relative millisecond timeouts are represented as ints in the code, and there should be no way for this calculation to overflow since the config properties to specify the backoff are limited to 60*60*1000 milliseconds.